src/c.cpp

```diff
@@ -0,0 +1,358 @@
+    opentracing_tracer_t* tracer,
+    const char* operation_name)
+{
+    std::unique_ptr<opentracing_span_t> span(new opentracing_span_t());
+    span->impl_ = tracer->impl_->StartSpan(operation_name).release();
+    return span.release();
+}
+
+extern "C" opentracing_span_t* opentracing_start_span_with_options(
+    opentracing_tracer_t* tracer,
+    const char* operation_name,
+    const opentracing_startspanoptions_t* options)
+{
+    opentracing::StartSpanOptions cppOptions;
+    cppOptions.start_steady_timestamp =
+        timespecToTimePoint<opentracing::SteadyClock>(
+            options->start_steady_timestamp);
+    cppOptions.start_system_timestamp =
+        timespecToTimePoint<opentracing::SystemClock>(
+            options->start_system_timestamp);
+
+    for (auto i = 0; i < options->num_references; ++i) {
+        const auto* ref = options->references[i];
+        cppOptions.references.emplace_back(
+                static_cast<opentracing::SpanReferenceType>(ref->type),
+                ref->context->impl_);
+    }
+
+    for (auto i = 0; i < options->num_tags; ++i) {
+        const auto* tag = options->tags[i];
+        cppOptions.tags.emplace_back(tag->key, tag->value->impl_);
+    }
+
+    std::unique_ptr<opentracing_span_t> span(new opentracing_span_t());
+    span->impl_ = tracer->impl_->StartSpanWithOptions(
+                    operation_name, cppOptions).release();
+    return span.release();
+}
+
+extern "C" void opentracing_free_span(opentracing_span_t* span)
+{
+    delete span;
+}
+
+extern "C" int opentracing_inject_binary(
+    const opentracing_tracer_t* tracer,
+    const opentracing_spancontext_t* sc,
+    opentracing_buffer_t* buffer)
+{
+    std::ostringstream oss;
+    try {
+        const auto result = tracer->impl_->Inject(*sc->impl_, oss);
+        if (!result) {
+            return result.error().value();
+        }
+
+        const auto dataStr = oss.str();
+        const auto dataStrSize = dataStr.size();
+        buffer->data = static_cast<char*>(malloc(dataStrSize));
+        if (!buffer->data) {
+            buffer->size = 0;
+            return -1;
+        }
+
+        buffer->size = static_cast<int>(oss.str().size());
+        std::memcpy(buffer->data, dataStr.c_str(), dataStrSize);
+        return 0;
+    } catch (...) {
+        return -1;
+    }
+}
+
+extern "C" int opentracing_extract_binary(
+    const opentracing_tracer_t* tracer,
+    opentracing_spancontext_t* sc,
+    const opentracing_buffer_t* buffer)
+{
+    try {
+        std::istringstream iss(
+            opentracing::string_view(
+                buffer->data, static_cast<size_t>(buffer->size)));
+        auto result = tracer->impl_->Extract(iss);
+        if (!result) {
+            return result.error().value();
+        }
+        sc->impl_ = result->release();
+        return 0;
+    } catch (...) {
+        return -1;
+    }
+}
+
+extern "C" void opentracing_close_tracer(opentracing_tracer_t* tracer)
```

Comment:
Shouldn't there also be an `opentracing_free_tracer` function? `Tracer::Close` may wait for spans to flush, so it's not always desirable for applications to call it when freeing resources since it could make the program take much longer to terminate.

Reply:
Interesting distinction. I can add free separately.